Xtensa Support

b1llow opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [ ] I've documented or updated the documentation of every API function and struct this PR changes.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

...

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Or what test cases you added. Disassembly tests should be added in suite/cs_test/issues.cs -->

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

Rot127 commented on arch/Xtensa/XtensaGenCSRegEnum.inc:
Please don't forget to delete these files, once they are included directly into the `xtensa.h` header.

Rot127 commented on suite/auto-sync/src/autosync/cpptranslator/patches/GetOperandRegImm.py:
```diff
@@ -12,7 +12,7 @@ class GetOperandRegImm(Patch):
     to      MCOperand_getReg(OPERAND)
 
-    Same for isImm()
+    Same for getImm()|getExpr
```
So far I excluded support for expressions from Capstone. Because they were used for symbols, which is out of scope for Capstone. Is this different in Xtensa?

Rot127 reviewed (commented):
Please keep in mind the `LITBASE` mechanic Xtensa has.
No other disassembler seems to support it. So it would be nice if CS gives at least an option to take care of it.
See: https://youtu.be/QMO2vMBcx7Y?feature=shared&t=808

XVilka commented:
@imbillow what LLVM version you have used as a base for this PR? Mainstream LLVM 18? Or a fork from developers working on the Xtensa patches?

b1llow commented:
> @imbillow what LLVM version you have used as a base for this PR? Mainstream LLVM 18? Or a fork from developers working on the Xtensa patches?

I'm using the auto-sync branch of llvm-capstone, I think it's llvm-18, I'm not sure.

```
commit 5943ec6923d64e63b9645aa230cd9b86bd63a51b (HEAD -> auto-sync, origin/auto-sync, origin/HEAD)
Author: Rot127 <unisono@quyllur.org>
Date:   Tue Jun 4 03:29:20 2024 -0500

    Add Alpha and LoongArch to the CI tests.
```

Rot127 commented:
It is LLVM 18. Please quickly diff the `Target/Xtensa` directories against each other. If there are vast differences, we can consider merging them earlier.

Rot127 commented on suite/cstest/Makefile:
```diff
@@ -1,12 +1,12 @@
 INCLUDE = include ../../include
 BUILD = build
-LIBRARY = -lcmocka -lcapstone -L../..
+LIBRARY = -lcapstone -L../..
```
Just saw this. Better rebase on the AArch64 PR (https://github.com/capstone-engine/capstone/pull/2298). It makes some changes to the auto-sync scripts and also makes the cstest build more conveninet (just pass `-DCAPSTONE_BUILD_CSTEST` when building capstone)

Rot127 replied:
So these things can be removed.

b1llow commented:
A build based on the latest next branch still fails tests, but I'm seeing different results from the CI. I wonder if this is m2 macOS-specific behavior.

@Rot127

```
➤ git rev-parse --verify HEAD
60d5b7ec2f62e0115cb0833e6429fb9057f5867a
➤ cmake --build cmake-build-debug/
ninja: no work to do.
➤ cmake --install cmake-build-debug/ --prefix='.local'
...
➤ ./.local/bin/cstest -f tests/cs_details/issue.cs 
[+] TARGET: tests/cs_details/issue.cs
[==========] Testing issues: Running 51 test(s).
[ RUN      ] !# issue 0 ARM operand groups 0x90,0xe8,0x0e,0x00 == ldm.w r0, {r1, r2, r3} ;
[       OK ] !# issue 0 ARM operand groups 0x90,0xe8,0x0e,0x00 == ldm.w r0, {r1, r2, r3} ;
[ RUN      ] !# issue 0 ARM operand groups 0x0e,0xc8 == ldm r0!, {r1, r2, r3} ;
[       OK ] !# issue 0 ARM operand groups 0x0e,0xc8 == ldm r0!, {r1, r2, r3} ;
[ RUN      ] !# issue 0 ARM operand groups 0x00,0x2a,0xf7,0xee == vmov.f32 s5, #1.000000e+00 ;
[       OK ] !# issue 0 ARM operand groups 0x00,0x2a,0xf7,0xee == vmov.f32 s5, #1.000000e+00 ;
[ RUN      ] !# issue 0 ARM operand groups 0x0f,0x00,0x71,0xe3 == cmn r1, #15 ;
[       OK ] !# issue 0 ARM operand groups 0x0f,0x00,0x71,0xe3 == cmn r1, #15 ;
[ RUN      ] !# issue 0 ARM operand groups 0x03,0x20,0xb0,0xe1 == movs r2, r3 ;
[       OK ] !# issue 0 ARM operand groups 0x03,0x20,0xb0,0xe1 == movs r2, r3 ;
[ RUN      ] !# issue 0 ARM operand groups 0xfd,0x8f == ldrh r5, [r7, #62] ;
[       OK ] !# issue 0 ARM operand groups 0xfd,0x8f == ldrh r5, [r7, #62] ;
[ RUN      ] !# issue 0 ARM operand groups 0x61,0xb6 == cpsie f ;
[       OK ] !# issue 0 ARM operand groups 0x61,0xb6 == cpsie f ;
[ RUN      ] !# issue 0 ARM operand groups 0x18,0xf8,0x03,0x1e == ldrbt r1, [r8, #3] ;
[       OK ] !# issue 0 ARM operand groups 0x18,0xf8,0x03,0x1e == ldrbt r1, [r8, #3] ;
[ RUN      ] !# issue 0 ARM operand groups 0xb0,0xf8,0x01,0xf1 == pldw [r0, #257] ;
[       OK ] !# issue 0 ARM operand groups 0xb0,0xf8,0x01,0xf1 == pldw [r0, #257] ;
[ RUN      ] !# issue 0 ARM operand groups 0xd3,0xe8,0x08,0xf0 == tbb [r3, r8] ;
[       OK ] !# issue 0 ARM operand groups 0xd3,0xe8,0x08,0xf0 == tbb [r3, r8] ;
[ RUN      ] !# issue 0 ARM operand groups 0xd3,0xe8,0x18,0xf0 == tbh [r3, r8, lsl #1] ;
[       OK ] !# issue 0 ARM operand groups 0xd3,0xe8,0x18,0xf0 == tbh [r3, r8, lsl #1] ;
[ RUN      ] !# issue 0 ARM operand groups 0xaf,0xf3,0x43,0x85 == cpsie i, #3 ;
[       OK ] !# issue 0 ARM operand groups 0xaf,0xf3,0x43,0x85 == cpsie i, #3 ;
[ RUN      ] !# issue 0 ARM operand groups 0xbf,0xf3,0x6f,0x8f == isb sy ;
[       OK ] !# issue 0 ARM operand groups 0xbf,0xf3,0x6f,0x8f == isb sy ;
[ RUN      ] !# issue 0 ARM operand groups 0x59,0xea,0x7b,0x89 == csel r9, r9, r11, vc ;
[       OK ] !# issue 0 ARM operand groups 0x59,0xea,0x7b,0x89 == csel r9, r9, r11, vc ;
[ RUN      ] !# issue 0 ARM operand groups 0xbf,0xf3,0x56,0x8f == dmb nshst ;
[       OK ] !# issue 0 ARM operand groups 0xbf,0xf3,0x56,0x8f == dmb nshst ;
[ RUN      ] !# issue 0 ARM operand groups 0x31,0xfa,0x02,0xf2 == lsrs.w r2, r1, r2 ;
[       OK ] !# issue 0 ARM operand groups 0x31,0xfa,0x02,0xf2 == lsrs.w r2, r1, r2 ;
[ RUN      ] !# issue 0 ARM operand groups 0x5f,0xf0,0x0c,0x01 == movseq.w r1, #12 ;
[       OK ] !# issue 0 ARM operand groups 0x5f,0xf0,0x0c,0x01 == movseq.w r1, #12 ;
[ RUN      ] !# issue 0 ARM operand groups 0x52,0xe8,0x01,0x1f == ldrex r1, [r2, #4] ;
[       OK ] !# issue 0 ARM operand groups 0x52,0xe8,0x01,0x1f == ldrex r1, [r2, #4] ;
[ RUN      ] !# issue 0 ARM operand groups 0xdf,0xec,0x1d,0x1a == vscclrmhi {s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15, s16, s17, s18, s19, s20, s21, s22, s23, s24, s25, s26, s27, s28, s29, s30, s31, vpr} ;
[       OK ] !# issue 0 ARM operand groups 0xdf,0xec,0x1d,0x1a == vscclrmhi {s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15, s16, s17, s18, s19, s20, s21, s22, s23, s24, s25, s26, s27, s28, s29, s30, s31, vpr} ;
[ RUN      ] !# issue 0 ARM operand groups 0x9f,0xec,0x06,0x5b == vscclrm {d5, d6, d7, vpr} ;
[       OK ] !# issue 0 ARM operand groups 0x9f,0xec,0x06,0x5b == vscclrm {d5, d6, d7, vpr} ;
[ RUN      ] !# issue 0 ARM operand groups 0xbc,0xfd,0x7f,0xaf == vldrh.u32 q5, [r4, #254]! ;
[       OK ] !# issue 0 ARM operand groups 0xbc,0xfd,0x7f,0xaf == vldrh.u32 q5, [r4, #254]! ;
[ RUN      ] !# issue 0 ARM operand groups 0x80,0xfc,0x80,0x1e == vst20.16 {q0, q1}, [r0] ;
[       OK ] !# issue 0 ARM operand groups 0x80,0xfc,0x80,0x1e == vst20.16 {q0, q1}, [r0] ;
[ RUN      ] !# issue 0 ARM operand groups 0x98,0xfc,0x4e,0x08 == vcadd.f32 q0, q4, q7, #90 ;
[       OK ] !# issue 0 ARM operand groups 0x98,0xfc,0x4e,0x08 == vcadd.f32 q0, q4, q7, #90 ;
[ RUN      ] !# issue 0 ARM operand groups 0x94,0xfd,0x46,0x48 == vcadd.f32 q2, q2, q3, #270 ;
[       OK ] !# issue 0 ARM operand groups 0x94,0xfd,0x46,0x48 == vcadd.f32 q2, q2, q3, #270 ;
[ RUN      ] !# issue 0 ARM operand groups 0x9d,0xec,0x82,0x6e == vldrb.s16 q3, [sp, q1] ;
[       OK ] !# issue 0 ARM operand groups 0x9d,0xec,0x82,0x6e == vldrb.s16 q3, [sp, q1] ;
[ RUN      ] !# issue 0 ARM operand groups 0x90,0xec,0x12,0x6f == vldrh.s32 q3, [r0, q1] ;
[       OK ] !# issue 0 ARM operand groups 0x90,0xec,0x12,0x6f == vldrh.s32 q3, [r0, q1] ;
[ RUN      ] !# issue 0 ARM operand groups 0x5f,0xea,0x2d,0x83 == sqrshrl lr, r3, #64, r8 ;
[       OK ] !# issue 0 ARM operand groups 0x5f,0xea,0x2d,0x83 == sqrshrl lr, r3, #64, r8 ;
[ RUN      ] !# issue 0 ARM operand groups 0x82,0xfd,0x21,0xff == vstrd.64 q7, [q1, #264] ;
[       OK ] !# issue 0 ARM operand groups 0x82,0xfd,0x21,0xff == vstrd.64 q7, [q1, #264] ;
[ RUN      ] !# issue 0 ARM operand groups 0x06,0x16,0x72,0xe6 == ldrbt r1, [r2], -r6, lsl #12 ;
[       OK ] !# issue 0 ARM operand groups 0x06,0x16,0x72,0xe6 == ldrbt r1, [r2], -r6, lsl #12 ;
[ RUN      ] !# issue 0 ARM operand groups 0xf6,0x50,0x33,0xe1 == ldrsh r5, [r3, -r6]! ;
[       OK ] !# issue 0 ARM operand groups 0xf6,0x50,0x33,0xe1 == ldrsh r5, [r3, -r6]! ;
[ RUN      ] !# issue 0 ARM operand groups 0x1e,0x19,0x7a,0xfd == ldc2l p9, c1, [r10, #-120]! ;
[       OK ] !# issue 0 ARM operand groups 0x1e,0x19,0x7a,0xfd == ldc2l p9, c1, [r10, #-120]! ;
[ RUN      ] !# issue 0 ARM operand groups 0x12,0x31,0x7c,0xfc == ldc2l p1, c3, [r12], #-72 ;
[       OK ] !# issue 0 ARM operand groups 0x12,0x31,0x7c,0xfc == ldc2l p1, c3, [r12], #-72 ;
[ RUN      ] !# issue 0 ARM operand groups 0xa4,0xf9,0x6d,0x0e == vld3.16 {d0[], d2[], d4[]}, [r4]! ;
[  ERROR   ] --- 0xa4,0xf9,0x6d,0x0e --- "operands[3].mem.index: REG = r4" not in "vld3.16 {d0[], d2[], d4[]}, [r4]! ; op_count: 4 ; operands[0].type: REG = d0 ; operands[0].access: WRITE ; operands[1].type: REG = d2 ; operands[1].access: WRITE ; operands[2].type: REG = d4 ; operands[2].access: WRITE ; operands[3].type: MEM ; operands[3].mem.base: REG = r4 ; operands[3].mem.scale: 0 ; operands[3].access: READ | WRITE ; Write-back: True ; Vector-size: 16 ; Registers read: r4 ; Registers modified: r4 d0 d2 d4 ; Groups: HasNEON ;"
[  ERROR   ] --- [   LINE   ] --- /Users/aya/Source/work/cs/suite/cstest/src/capstone_test.c:287: error: Failure!
[  FAILED  ] !# issue 0 ARM operand groups 0xa4,0xf9,0x6d,0x0e == vld3.16 {d0[], d2[], d4[]}, [r4]! ;
[ RUN      ] !# issue 0 ARM operand groups 0x0d,0x50,0x66,0xe4 == strbt r5, [r6], #-13 ;
[       OK ] !# issue 0 ARM operand groups 0x0d,0x50,0x66,0xe4 == strbt r5, [r6], #-13 ;
[ RUN      ] !# issue 0 ARM operand groups 0x00,0x10,0x4f,0xe2 == sub r1, pc, #0 ;
[       OK ] !# issue 0 ARM operand groups 0x00,0x10,0x4f,0xe2 == sub r1, pc, #0 ;
[ RUN      ] !# issue 0 ARM operand groups 0x9f,0x51,0xd3,0xe7 == bfc r5, #3, #17 ;
[       OK ] !# issue 0 ARM operand groups 0x9f,0x51,0xd3,0xe7 == bfc r5, #3, #17 ;
[ RUN      ] !# issue 0 ARM operand groups 0xd8,0xe8,0xff,0x67 == ldaexd r6, r7, [r8] ;
[       OK ] !# issue 0 ARM operand groups 0xd8,0xe8,0xff,0x67 == ldaexd r6, r7, [r8] ;
[ RUN      ] !# issue 0 ARM operand groups 0x30,0x0f,0xa6,0xe6 == ssat16 r0, #7, r0 ;
[       OK ] !# issue 0 ARM operand groups 0x30,0x0f,0xa6,0xe6 == ssat16 r0, #7, r0 ;
[ RUN      ] !# issue 0 ARM operand groups 0x9a,0x8f,0xa0,0xe6 == ssat r8, #1, r10, lsl #31 ;
[       OK ] !# issue 0 ARM operand groups 0x9a,0x8f,0xa0,0xe6 == ssat r8, #1, r10, lsl #31 ;
[ RUN      ] !# issue 0 ARM operand groups 0x40,0x1b,0xf5,0xee == vcmp.f64 d17, #0 ;
[       OK ] !# issue 0 ARM operand groups 0x40,0x1b,0xf5,0xee == vcmp.f64 d17, #0 ;
[ RUN      ] !# issue 0 ARM operand groups 0x05,0xf0,0x2f,0xe3 == msr CPSR_fsxc, #5 ;
[       OK ] !# issue 0 ARM operand groups 0x05,0xf0,0x2f,0xe3 == msr CPSR_fsxc, #5 ;
[ RUN      ] !# issue 0 ARM operand groups 0xa4,0xf9,0xed,0x0b == vld4.32 {d0[1], d2[1], d4[1], d6[1]}, [r4:128]! ;
[  ERROR   ] --- 0xa4,0xf9,0xed,0x0b --- "operands[4].mem.index: REG = r4" not in "vld4.32 {d0[1], d2[1], d4[1], d6[1]}, [r4:0x80]! ; op_count: 5 ; operands[0].type: REG = d0 ; operands[0].neon_lane = 1 ; operands[0].access: READ | WRITE ; operands[1].type: REG = d2 ; operands[1].neon_lane = 1 ; operands[1].access: READ | WRITE ; operands[2].type: REG = d4 ; operands[2].neon_lane = 1 ; operands[2].access: READ | WRITE ; operands[3].type: REG = d6 ; operands[3].neon_lane = 1 ; operands[3].access: READ | WRITE ; operands[4].type: MEM ; operands[4].mem.base: REG = r4 ; operands[4].mem.scale: 0 ; operands[4].access: READ | WRITE ; Write-back: True ; Vector-size: 32 ; Registers read: d0 d2 d4 d6 r4 ; Registers modified: r4 d0 d2 d4 d6 ; Groups: HasNEON ;"
[  ERROR   ] --- [   LINE   ] --- /Users/aya/Source/work/cs/suite/cstest/src/capstone_test.c:287: error: Failure!
[  FAILED  ] !# issue 0 ARM operand groups 0xa4,0xf9,0xed,0x0b == vld4.32 {d0[1], d2[1], d4[1], d6[1]}, [r4:128]! ;
[ RUN      ] !# issue 0 ARM operand groups 0x42,0x03,0xb0,0xf3 == aesd.8 q0, q1 ;
[       OK ] !# issue 0 ARM operand groups 0x42,0x03,0xb0,0xf3 == aesd.8 q0, q1 ;
[ RUN      ] !# issue 0 ARM operand groups 0x11,0x57,0x54,0xfc == mrrc2 p7, #1, r5, r4, c1 ;
[       OK ] !# issue 0 ARM operand groups 0x11,0x57,0x54,0xfc == mrrc2 p7, #1, r5, r4, c1 ;
[ RUN      ] !# issue 0 ARM operand groups 0xd3,0x2f,0x82,0xe6 == pkhtb r2, r2, r3, asr #31 ;
[       OK ] !# issue 0 ARM operand groups 0xd3,0x2f,0x82,0xe6 == pkhtb r2, r2, r3, asr #31 ;
[ RUN      ] !# issue 0 ARM operand groups 0x93,0x27,0x82,0xe6 == pkhbt r2, r2, r3, lsl #15 ;
[       OK ] !# issue 0 ARM operand groups 0x93,0x27,0x82,0xe6 == pkhbt r2, r2, r3, lsl #15 ;
[ RUN      ] !# issue 0 ARM operand groups 0xb4,0x10,0xf0,0xe0 == ldrht r1, [r0], #4 ;
[       OK ] !# issue 0 ARM operand groups 0xb4,0x10,0xf0,0xe0 == ldrht r1, [r0], #4 ;
[ RUN      ] !# issue 0 ARM operand groups 0x2f,0xfa,0xa1,0xf3 == sxtb16 r3, r1, ror #16 ;
[       OK ] !# issue 0 ARM operand groups 0x2f,0xfa,0xa1,0xf3 == sxtb16 r3, r1, ror #16 ;
[ RUN      ] !# issue 0 ARM operand groups 0x00,0x02,0x01,0xf1 == setend be ;
[       OK ] !# issue 0 ARM operand groups 0x00,0x02,0x01,0xf1 == setend be ;
[ RUN      ] !# issue 0 ARM operand groups 0xd0,0xe8,0xaf,0x0f == lda r0, [r0]
[       OK ] !# issue 0 ARM operand groups 0xd0,0xe8,0xaf,0x0f == lda r0, [r0]
[ RUN      ] !# issue 0 ARM operand groups 0xef,0xf3,0x11,0x85 == ldrhi pc, [r1, #-0x3ef]
[  ERROR   ] --- 0xef,0xf3,0x11,0x85 --- "Groups: IsARM" not in "ldrhi pc, [r1, #-0x3ef] ; op_count: 2 ; operands[0].type: REG = r15 ; operands[0].access: WRITE ; operands[1].type: MEM ; operands[1].mem.base: REG = r1 ; operands[1].mem.scale: 0 ; operands[1].mem.disp: 0x3ef ; operands[1].access: READ ; Subtracted: True ; Code condition: 8 ; Registers read: cpsr r1 ; Registers modified: r15 ; Groups: IsARM jump ;"
[  ERROR   ] --- [   LINE   ] --- /Users/aya/Source/work/cs/suite/cstest/src/capstone_test.c:287: error: Failure!
[  FAILED  ] !# issue 0 ARM operand groups 0xef,0xf3,0x11,0x85 == ldrhi pc, [r1, #-0x3ef]
[==========] Testing issues: 51 test(s) run.
[  PASSED  ] 48 test(s).
[  FAILED  ] Testing issues: 3 test(s), listed below:
[  FAILED  ] !# issue 0 ARM operand groups 0xa4,0xf9,0x6d,0x0e == vld3.16 {d0[], d2[], d4[]}, [r4]! ;
[  FAILED  ] !# issue 0 ARM operand groups 0xa4,0xf9,0xed,0x0b == vld4.32 {d0[1], d2[1], d4[1], d6[1]}, [r4:128]! ;
[  FAILED  ] !# issue 0 ARM operand groups 0xef,0xf3,0x11,0x85 == ldrhi pc, [r1, #-0x3ef]

 3 FAILED TEST(S)
[+] DONE: tests/cs_details/issue.cs
[!] Noted:
[  ERROR   ] --- "<capstone result>" != "<user result>"
```

Rot127 commented:
The `tests/cs_details/issue.cs` are not tested in the current `next`. Added them though in the in ASAN PR. But they are obsolete with the AArch64 PR anyways. There they are changed again.

XVilka commented on .github/workflows/CITest.yml:
This change is probably better to extract into a separate PR

XVilka replied:
Nevermind, it was imported commit.

XVilka commented on HACK.TXT:
Add also yourself to CREDITS

XVilka commented on bindings/python/test_xtensa.py:
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python
+
+# Capstone Python bindings, by Nguyen Anh Quynnh <aquynh@gmail.com>
```
Should it be your copyright? Also, those are not bindings, just a test

XVilka commented:
@imbillow note also, Xtensa could be both little endian and big endian.

https://0x04.net/~mwk/doc/xtensa.pdf

aquynh commented:
thank you for this amazing effort!

few quick comments:

- please add this new arch to README
- please add a test file into tests/
- please use capitalized letters for all enums, for example Xtensa_REG_SAR should be XTENSA_REG_SAR

b1llow commented:
> thank you for this amazing effort!
> 
> few quick comments:
> 
> * please add this new arch to README
> * please add a test file into tests/
> * please use capitalized letters for all enums, for example Xtensa_REG_SAR should be XTENSA_REG_SAR

The names of these enum types are generated from llvm-capstone via the auto-sync script, and I think it's best not to change them for case.

b1llow commented:
> @imbillow note also, Xtensa could be both little endian and big endian.
> 
> [0x04.net/~mwk/doc/xtensa.pdf](https://0x04.net/~mwk/doc/xtensa.pdf)

https://github.com/llvm/llvm-project/blob/5021e6dd548323e1169be3d466d440009e6d1f8e/llvm/lib/Target/Xtensa/Disassembler/XtensaDisassembler.cpp#L256

The latest version of LLVM also does not support big-endian mode, probably because LLVM's xtensa support is still experimental.

Although there is a third party repository that supports xtensa 
https://github.com/espressif/llvm-project

XVilka commented:
> > @imbillow note also, Xtensa could be both little endian and big endian.
> > [0x04.net/~mwk/doc/xtensa.pdf](https://0x04.net/~mwk/doc/xtensa.pdf)
> 
> https://github.com/llvm/llvm-project/blob/5021e6dd548323e1169be3d466d440009e6d1f8e/llvm/lib/Target/Xtensa/Disassembler/XtensaDisassembler.cpp#L256
> 
> The latest version of LLVM also does not support big-endian mode, probably because LLVM's xtensa support is still experimental.
> 
> Although there is a third party repository that supports xtensa https://github.com/espressif/llvm-project

At least please mark the Xtensa mode as LE then, until BE is added, so that the mode enum change will not be necessary in the future.

Rot127 commented:
@imbillow Enums get fixed with: https://github.com/capstone-engine/llvm-capstone/pull/50

aquynh commented:
> > thank you for this amazing effort!
> > few quick comments:
> > 
> > * please add this new arch to README
> > * please add a test file into tests/
> > * please use capitalized letters for all enums, for example Xtensa_REG_SAR should be XTENSA_REG_SAR
> 
> The names of these enum types are generated from llvm-capstone via the auto-sync script, and I think it's best not to change them for case.

please change the case, as all other arcs are doing that. 

the auto-sync code should update to have this, too.

Rot127 commented:
@imbillow You can use the latest `main` branch of the `llvm-capstone`. It will generate the enums in capital letters.

b1llow committed:
Xtensa: add xtensa support

b1llow committed:
Xtensa fixes

- fix MCExpr
- fix Xtensa_add_cs_detail
- add `add_cs_detail`
- add `MCExpr *MCOperand_getExpr(const MCOperand *MC)` `void printExpr(const MCExpr *E, SStream *O)`

autosync fix

- fix StreamOperation.py
- replace `report_fatal_error` with `CS_ASSERT`
- fix patch StreamOperation.py
- replace `assert` with `CS_ASSERT`
- fix AddCSDetail.py
- fix QualifiedIdentifier

b1llow commented:
@aquynh @kabeor @Rot127 this PR is ready to be merged, please check again.

Rot127 commented:
Please check the `clang-tidy` warning:
```
 /home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:75:29: note: The left operand of '==' is a garbage value due to array index out of bounds
                if (SRDecoderTable[i + 1] == RegNo) {
                    ~~~~~~~~~~~~~~~~~~~~~ ^
```

This can be out of bounds

Rot127 commented on arch/Xtensa/priv.h:
```diff
@@ -0,0 +1,17 @@
+#define CAPSTONE_PRIV_H
+
+#define report_fatal_error(x) assert(0 && x)
+#define llvm_unreachable(x) assert(0 && x)
```
`llvm-unreachable` should be replaced by the auto-sync scripts. Or are they not?

b1llow replied:
Not sure why they haven't been replaced.

b1llow replied:
I think I know why, because *.inc files are not translated.

Rot127 replied:
Ah yes. They are not. But in general it is fine. Just add a function to `MCInst` or somewhere where it fits well.
Just use `CS_ASSERT` in there. not the normal one please.

Rot127 commented on arch/Xtensa/priv.h:
```diff
@@ -0,0 +1,17 @@
+#define report_fatal_error(x) assert(0 && x)
+#define llvm_unreachable(x) assert(0 && x)
+#define MCRegister unsigned
+#define MCExpr void
```
Move them as `typedef` to `MCRegisterInfo.h` please

Rot127 commented on arch/Xtensa/priv.h:
```diff
@@ -0,0 +1,17 @@
+#define llvm_unreachable(x) assert(0 && x)
+#define MCRegister unsigned
+#define MCExpr void
+#define MCOperand_getExpr(MC) assert(0 && "unimplemented expr")
```
This one as normal function to `MCInst.h`/`MCInst.c` please.

Rot127 commented on arch/Xtensa/priv.h:
```diff
@@ -0,0 +1,17 @@
+#define MCRegister unsigned
+#define MCExpr void
+#define MCOperand_getExpr(MC) assert(0 && "unimplemented expr")
+#define printExpr(E, O) assert(0 && "unimplemented expr")
```
This one to `MCInstPrinter.c` also as normal function.

b1llow replied:
`static void printExpr(const MCExpr *Expr, raw_ostream &OS)` is a static function and it is actually defined in XtensaInstPrinter.cpp

Rot127 replied:
Ok, then it is fine. I didn't check the `cpp` source.

Rot127 commented on arch/Xtensa/XtensaMapping.c:
```diff
@@ -0,0 +1,206 @@
+	}
+
+	const map_insn_ops *ops = &insn_operands[opcode];
+	cs_xtensa *detail = &MI->flat_insn->detail->xtensa;
```
```suggestion
	cs_xtensa *detail = Xtensa_get_detail(MI);
```
Requires of course to add it in `Mapping.h`

Rot127 commented on arch/Xtensa/XtensaMapping.c:
```diff
@@ -0,0 +1,206 @@
+
+		detail->op_count++;
+		operand++;
+	}
```
This logic doesn't work, unfortunately.

The reason all the `add_cs_detail()` calls are patched into the `ARCHInstPrinter.c` is because `MCOperand->ImmVal` and `MCOperand->RegVal` are not always the actual values printed in the asm-text. And with it, not the value saved to `cs_detail`.

A lot of times some operations are done on the raw values extracted from the bytes.

E.g.:
https://github.com/imbillow/capstone/blob/6b8e7bc7601e3f89aa0e8dc4f4cf2bc560ae4068/arch/Xtensa/XtensaInstPrinter.c#L115-L122
https://github.com/imbillow/capstone/blob/6b8e7bc7601e3f89aa0e8dc4f4cf2bc560ae4068/arch/Xtensa/XtensaInstPrinter.c#L101

But the operation performed on the raw values depend on the operand type. At this point in the code the operand type information is lost. So you cannot determine, what operation is performed on it.

Hence, the patched in `add_cs_detail`.

Please patch the `add_cs_detail()` calls again and add detail tests in `issues.cs`. So they cover all operand types in `XtensaInstPrinter`. The operand values in `cs_xtensa` should match the values in the asm-text.

Rot127 commented on arch/Xtensa/XtensaInstPrinter.c:
```diff
@@ -0,0 +1,279 @@
+		case 256:
+			break;
+		default:
+			break;
```
You cannot remove the `assert` at this place. Please update the `patches/Assert.py` patch and replace `asserts` with `CS_ASSERT()`.

Rot127 replied:
Also above

Rot127 replied:
You can also do it for the `llvm_unreachable`

Rot127 commented on arch/Xtensa/XtensaMapping.c:
```diff
@@ -0,0 +1,206 @@
+#include <capstone/xtensa.h>
```
Copyright missing

Rot127 commented on cstool/cstool_xtensa.inc:
```diff
@@ -0,0 +1,62 @@
+#include <stdio.h>
```
Please add your copyright. Also, why an `inc` file? It tends to break syntax highlighting and language servers.

b1llow replied:
Because I feel like this print insn detail logic is repeated multiple times (cstool_xxx.c, cstest, test_xxx.c, test_xxx.py), and I'm just a little resistant to copying it four times?

Rot127 replied:
This is true. But the test files will be gone very soon. I currently rewrite testing completely. So only the cstool will stay there. You can wait for it, it shouldn't take long.

Rot127 replied:
You can move the code now `cstool_xtensa`.

Rot127 commented on cstool/cstool.h:
```diff
@@ -22,5 +22,6 @@ void print_insn_detail_tricore(csh handle, cs_insn *ins);
 void print_insn_detail_alpha(csh handle, cs_insn *ins);
 void print_insn_detail_hppa(csh handle, cs_insn *ins);
 void print_insn_detail_loongarch(csh handle, cs_insn *ins);
+#include "cstool_xtensa.inc"
```
As above. Please implement it as the others archs do it.

Rot127 commented on include/capstone/capstone.h:
```diff
@@ -243,6 +244,7 @@ typedef enum cs_mode {
 	CS_MODE_HPPA_20W = CS_MODE_HPPA_20 | (1 << 3), ///< HPPA 2.0 wide
 	CS_MODE_LOONGARCH32  = 1 << 0,        ///< LoongArch32
 	CS_MODE_LOONGARCH64  = 1 << 1,        ///< LoongArch64
+	CS_MODE_XTENSA = 1 << 1, ///< Xtensa
```
```suggestion
	CS_MODE_XTENSA = 1 << 0, ///< Xtensa
```

Rot127 reviewed (changes requested):
It seems like you fixed a lot by hand and didn't let the Auto-Sync patches do their work (at least for the things below). Was there a particular reason for it?

If you find a certain patch doesn't make sense for Xtensa at all, you can disable it with the `SomePatch.py::do_not_apply` (see `Patch.py`).

Also feel free to extend them where necessary. E.g. the `AddCSDetail` and the `Assert` patch.

Please add detail tests in `tests/cs_details/issue.cs`. You will see that they fail because the operand details don't match the ones in the asm text.

Although it might seem annoying to implement the `add_cs_detail` logic, it is the better bet in the long run. Please don't implement an alternative way of setting details.
If the Xtensa disassembler gets bigger and bigger with time or starts to use template functions for operand printing, it is likely to break again.

Also, I couldn't see any handling of the `LITBASE`. Do you already have an idea how to solve it?

XVilka commented:
@imbillow please rebase

b1llow commented:
I might revert the fuzz.yml change because it was completely turned off in #2396 that have been merged. And to fix it we need to change this file in another project: oss-fuzz/projects/capstone/Dockerfile

Rot127 commented:
Yeah, also I have to make changes to the fuzzer scripts as well with https://github.com/capstone-engine/capstone/pull/2384. So don't put too much effort into it.
You can also consider to rebase this one on top of https://github.com/capstone-engine/capstone/pull/2384.
I won't make changes to the module code there any longer. So you shouldn't get conflicts during rebasing it.

b1llow commented:
> Yeah, also I have to make changes to the fuzzer scripts as well with #2384. So don't put too much effort into it. You can also consider to rebase this one on top of #2384. I won't make changes to the module code there any longer. So you shouldn't get conflicts during rebasing it.

I feel tortured every time I look at your huge PR on chrome because it crashes the browser.

XVilka commented:
> > Yeah, also I have to make changes to the fuzzer scripts as well with #2384. So don't put too much effort into it. You can also consider to rebase this one on top of #2384. I won't make changes to the module code there any longer. So you shouldn't get conflicts during rebasing it.
> 
> I feel tortured every time I look at your huge PR on chrome because it crashes the browser.

You can try using `tig` or something like that. It would certainly not crash.

b1llow commented:
Currently compiling with the ninja generator on macos gives an error

```
cc -DVERSION_MAJOR=1 -DVERSION_MINOR=4 -DVERSION_PATCH=1 -DVERSION_DEVEL=0 -MMD -MP -I include  -std=c11 -Wall -Wextra -pedantic -Wconversion -Wwrite-strings -Wcast-align -Wpointer-arith -Winit-self -Wshadow -Wstrict-prototypes -Wmissing-prototypes -Wredundant-decls -Wundef -Wvla -Wdeclaration-after-statement -O2 -DNDEBUG -fPIC  -c -o build/release/shared/src/load.o src/load.c
src/load.c:24:10: fatal error: 'yaml.h' file not found
#include <yaml.h>
         ^~~~~~~~
1 error generated.
make: *** [build/release/shared/src/load.o] Error 1
ninja: build stopped: subcommand failed.
```

libyaml is also installed.

```
➤ pkg-config --cflags yaml-0.1
-I/opt/homebrew/Cellar/libyaml/0.2.5/include
```

Rot127 commented:
> I feel tortured every time I look at your huge PR on chrome because it crashes the browser.

Ouh yeah, sorry. There are a few thousand test files in there. Probably should be separated before merging.

> Currently compiling with the ninja generator on macos gives an error

Weird. Guess I didn't understand the `find_libary` documentation properly. Can you try this on please:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 21577ba42..3b376b18b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -887,7 +887,9 @@ endif()
 
 if(CAPSTONE_BUILD_CSTEST)
     include(ExternalProject)
-    find_library(libyaml yaml REQUIRED)
+    find_library(libyaml
+        NAMES libyaml yaml
+        REQUIRED)
     ExternalProject_Add(cmocka_ext
         PREFIX extern
         URL "https://cmocka.org/files/1.1/cmocka-1.1.7.tar.xz"
```

b1llow commented:
@Rot127

That would work but there are new problems.

```
capstone/suite/cstest/src/cstest.c:12:10: fatal error: 'threads.h' file not found
#include <threads.h>
         ^~~~~~~~~~~
1 error generated.
```

```
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -945,7 +945,7 @@ if(CAPSTONE_BUILD_CSTEST)
         DOWNLOAD_EXTRACT_TIMESTAMP true
         CONFIGURE_COMMAND ""
         BUILD_BYPRODUCTS ${LIBCYAML_LIB_DIR}/libcyaml.a
-        BUILD_COMMAND make
+        BUILD_COMMAND PKG_CONFIG=pkg-config make
         BUILD_IN_SOURCE true
         INSTALL_COMMAND ""
     )
```

I'm considering whether it's feasible to use rust for cstest.

Rot127 commented:
> I'm considering whether it's feasible to use rust for cstest.

I really did not want to make the building steps more complex. Especially because there is not much time.
Also it would put another layer of stuff in between.

And if we start using Rust, it would mean we might as well implement proper bindings for it. But the release is set to September.
Also, if we will implement Rust bindings in the future, we have to write a `rs-cstest` for them as well. So basically duplicating the work.

> That would work but there are new problems.

We can also live with global variables, although it hurts.
But cstest is not multi-threaded anyways (and doesn't need to be, it is fast enough this way)

```diff
diff --git a/suite/cstest/src/cstest.c b/suite/cstest/src/cstest.c
index a8793fc62..6c25f5db9 100644
--- a/suite/cstest/src/cstest.c
+++ b/suite/cstest/src/cstest.c
@@ -9,15 +9,14 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <threads.h>
 #include <time.h>
 
 // Pointer to the file list table
 // Must be a thread local, because we cannot pass arguments to `nftw`.
 // So the found test files can only be saved, very annoyingly,
 // to a global/thread-local mutable variables.
-thread_local char ***test_files = NULL;
-thread_local uint32_t file_count = 0;
+char ***test_files = NULL;
+uint32_t file_count = 0;
 
 static void help(const char *self)
 {
```

XVilka commented:
I guess this might be rebased one more time on top of the latest testing update, since it's now more stable and polished: https://github.com/capstone-engine/capstone/pull/2456

XVilka commented:
@imbillow could you please convert tests to YAML?

b1llow commented:
> @imbillow could you please convert tests to YAML?

Yes I can, I was just wondering if there was a way to quickly convert `.s.cs` files for testing into `yaml` form.

Rot127 commented:
@imbillow

I wrote a script for it.
I would need to restore my backup to get it though and it was only around 30-40 lines.
But for the few lines you've got there, it is probably easier to user multi-cursors in your editor.

But it is even better if you just generate the tests again. The `MCUpdater` is now rewritten. It produces better results as well.
~~Which LLVM repo do you use? I can check if everything works with it.~~
Ah, our one. Then we need a fix. They don't define Disassembler tests there. I take care of it tomorrow if you don't mind. You can also copy `llvm_root/llvm/test/MC/Xtensa/Core/` to `llvm_root/llvm/test/MC/Xtensa/Disassembler/` and run the updater.

Rot127 commented on CMakeLists.txt:
```diff
@@ -676,6 +679,23 @@ if (CAPSTONE_LOONGARCH_SUPPORT)
+        arch/Xtensa/XtensaMapping.h
+        arch/Xtensa/XtensaModule.h
+    )
+    set(TEST_SOURCES ${TEST_SOURCES} test_xtensa.c)
```
Delete the `test_xtensa` files (Python and C) and move the instruction tests to `tests/details/xtensa.yaml`

Rot127 commented on COMPILE_CMAKE.TXT:
If this gets merged after the documentation PR, you would need to add the changes to `BUILDING.md`

Rot127 commented on HACK.TXT:
Same here, this would go into `CONTRIBUTING.md` I think

Rot127 commented on MCInst.c:
```diff
@@ -165,6 +165,12 @@ void MCOperand_setFPImm(MCOperand *op, double Val)
+MCExpr *MCOperand_getExpr(const MCOperand *MC)
+{
+	assert(0 && "unimplemented expr");
+	return NULL;
```
It turned out that `SystemZ` and some `AArch64` instructions (I think) now also use the expressions. But emit the immediate value directly in LLVM, if symbol information is missing. With the SystemZ PR is will be:
```c
int64_t MCOperand_getExpr(const MCOperand *op)
{
	return op->ImmVal;
}
```

Rot127 commented on arch/Xtensa/XtensaDisassembler.c:
```diff
@@ -0,0 +1,302 @@
+static DecodeStatus decodeJumpOperand(MCInst *Inst, uint64_t Imm,
+				      int64_t Address, const void *Decoder)
+{
+	CS_ASSERT(CONCAT(isUInt, 18)(Imm) && "Invalid immediate");
```
```suggestion
	CS_ASSERT(isUIntN(18, Imm) && "Invalid immediate");
```

This function is provided in `MathExtras.h`. I added the patch for it also in the SystemZ PR.

Rot127 replied:
Please also use it everywhere else.

Rot127 replied:
I just see, that the documentation of the `isUInt` Patch is off:
```python
class IsUInt(Patch):
    """
    Patch   isUInt<N>(...)
    to      isUInt(..., N)  # Should be: "isUIntN(N, ...)"
    """
```

If you want to, you can fix it.

Rot127 commented on arch/Xtensa/XtensaDisassembler.c:
```diff
@@ -0,0 +1,302 @@
+		return MCDisassembler_Fail;
+	}
+
+	if (MODE_IS_BIG_ENDIAN(MI->csh->mode))
+		*Insn = (Bytes[2]) | (Bytes[1] << 8) |
+			((uint32_t)Bytes[0] << 16);
+	else
+		*Insn = (Bytes[2] << 16) | (Bytes[1] << 8) |
+			((uint32_t)Bytes[0]);
```
Can you move this please to `utils.c::readBytes24()`? There are already others of this kind.

Rot127 commented on arch/Xtensa/XtensaInstPrinter.c:
```diff
@@ -0,0 +1,313 @@
+	} else if (MCOperand_isExpr(MC))
+		MCExpr_print(MCOperand_getExpr(MC), O, NULL, true);
+	else
+		assert(0 && "Invalid operand");
```
A hard `assert` left.

Rot127 reviewed (changes requested):
I only came until `XtensaInstPrinter.c`. I'll do the rest tomorrow.

But looks good so far! Thanks for all the effort!

With the SystemZ, testing and all the other PRs, please rebase into https://github.com/capstone-engine/capstone/pull/2478

You get new `CS_ASSERT_RET`, `CS_ASSERT_RET_VAL` macros  there as well (just like in Rizin). Also it removes some invalid checks from the Auto-Sync workflow you no longer need to take care of.

Rot127 commented on arch/Xtensa/XtensaMapping.c:
```diff
@@ -0,0 +1,222 @@
+	}
+
+	const map_insn_ops *ops = insn_operands + MCInst_getOpcode(MI);
+	xop->access = (ops->ops + op_num)->access;
```
Please :)

```suggestion
	xop->access = map_get_op_access(MI, op_num);
```

Rot127 commented on arch/Xtensa/XtensaMapping.c:
```diff
@@ -0,0 +1,222 @@
+
+void Xtensa_add_cs_detail(MCInst *MI, xtensa_op_group op_group, va_list args)
+{
+	int op_num = va_arg(args, int);
```
`va_end(args);` missing

b1llow replied:
https://github.com/imbillow/capstone/blob/fef040d834a11b55a0c662b0b27fe35123c90594/arch/Xtensa/XtensaMapping.h#L29-L36

There is a va_end here, so no need for another?

Rot127 commented on arch/Xtensa/XtensaMapping.h:
```diff
@@ -0,0 +1,37 @@
+
+static inline void add_cs_detail(MCInst *MI, xtensa_op_group op_group, ...)
+{
+	if (!MI->flat_insn->detail)
```
```suggestion
	if (!detail_is_set(MI))
```
Rather important, because the function checks more than just `NULL`.

Rot127 commented on bindings/python/capstone/__init__.py:
```diff
@@ -467,16 +469,27 @@
+
+    if isinstance(_path, str):
+        return load_lib(_path)
+    elif isinstance(_path, ContextManager):
```
Out of curiosity. In which scenario is `_path` a `ContextManager`?

Rot127 commented on bindings/python/tests/test_xtensa.py:
Can be removed, since you moved the tests to the `yaml` files.

Rot127 commented on cstool/cstool.c:
```diff
@@ -333,6 +335,11 @@ static void usage(char *prog)
+		printf("        xtensa      Xtensa\n");
+		printf("        xtensabe    Xtensa + big endian\n");
+	}
+
```
Can be removed. It should be printed in the loop above.

Rot127 commented on arch/Xtensa/XtensaMapping.c:
```diff
@@ -0,0 +1,222 @@
+			MCOperand_getImm(MCInst_getOperand(MI, op_num + 1));
+		xop->type = XTENSA_OP_MEM;
+		xop->mem.base = reg;
+		xop->mem.disp = val;
```
`disp` is defined as `uint8_t`. Here you assign an `int64_t`. Can you please check the semantics and cast explicitly here if this is correct.

Rot127 commented on suite/auto-sync/src/autosync/ASUpdater.py:
```diff
@@ -54,13 +54,21 @@ def __init__(
-                USteps.MC,
-                USteps.PATCH_HEADER,
-            ]
+            if arch not in ["Xtensa"]:
```
Why no `MC` test for `Xtensa`? Is something not working?

b1llow replied:
It's been so long I've forgotten.

Rot127 commented on suite/auto-sync/src/autosync/MCUpdater.py:
```diff
@@ -466,10 +469,12 @@ def extract_llvm_mc_cmds(self, cmds: str) -> list[LLVM_MC_Command]:
         log.info("Check prerequisites")
         disas_tests = self.mc_dir.joinpath(f"Disassembler/{self.arch}")
-        test_paths = [disas_tests]
+        test_paths = [disas_tests, self.mc_dir.joinpath(self.arch)]
```
Mh, yeah. I just removed this, because the assembly tests break for a lot of more complicated archs. Because they allow to have multiple syntax inputs put only a single disassembler output. Which breaks stuff. So please add a check for `self.arch == "Xtensa` here.

Rot127 commented on suite/auto-sync/src/autosync/MCUpdater.py:
```diff
@@ -466,10 +469,12 @@ def extract_llvm_mc_cmds(self, cmds: str) -> list[LLVM_MC_Command]:
         log.info("Generate MC regression tests")
-        llvm_mc_cmds = self.run_llvm_lit(test_paths)
+        llvm_mc_cmds = self.run_llvm_lit(
+            [path for path in test_paths if exists_and_is_dir(path)]
```
What bug did you run into here?

b1llow replied:
It seems like there is an error when running a step using a path that doesn't exist.

Rot127 commented on suite/auto-sync/src/autosync/cpptranslator/patches/StreamOperation.py:
```diff
@@ -105,6 +105,37 @@ def get_patch(self, captures: [(Node, str)], src: bytes, **kwargs) -> bytes:
                 + last_op_text.replace(b"'", b'"')
                 + b");\n"
             )
+        elif last_op.type == "identifier":
```
This is really cool! Good idea!

Rot127 commented on suite/auto-sync/src/autosync/cpptranslator/patches/AddCSDetail.py:
```diff
@@ -70,14 +72,23 @@ def get_patch(self, captures: [(Node, str)], src: bytes, **kwargs) -> bytes:
         comp = captures[3][0]
         comp = get_text(src, comp.start_byte, comp.end_byte)
-        return b"void " + fcn_id + params + b"{ " + add_cs_detail + comp.strip(b"{")
+        return (
```
Good idea as well for now. In the long run the LLVM code will be it's complete own compilation unit and own interface. So we do not have all these annoying linking issues.

Rot127 reviewed (changes requested):
Nice! Awesome job! Also very useful additions to the patches!

Except the smaller requests above, two more things:

- `LITBASE`, sorry for being annoying about it. But a basic implementation is needed.
I think the following is enough for now and can be easily extended:
  - Add `size_t MCInst->LITBASE`, or something
  - Add a new option `CS_OPT_XTENSA_LITBASE`
  - Handle this option in `Xtensa_option`. The value given is the litbase address.
  - When adding the detail or printing the `L32Target` operand you check the `MI->LITBASE` if it is `0`. If `0` we handle it as if the `enable bit` is "unset". If not `0`, we treat it as `enable bit` "set".
  - This way the user gets relatively easy control over it, and we have a simple implementation for now.
  - Please add an integration test of it in `tests/integration/`
  - Reference: [4.3.3.2 The Literal Base Register](https://www.cadence.com/content/dam/cadence-www/global/en_US/documents/tools/silicon-solutions/compute-ip/isa-summary.pdf)
- Please add `Xtensa` to the fuzzer. Just add it to the other archs in `suite/fuzz/fuzz_diff.c`, `suite/fuzz/platform.c`, `suite/fuzz/fuzz_harness`, `suite/fuzz/drivermc`
- Also please update `docs/cs_v6_release_guide.md` with Xtensa and describe the LITBASE functionality there briefly.

b1llow committed:
Xtensa fix

b1llow committed:
Xtensa fix .py

b1llow committed:
add Xtensa to the fuzzer

b1llow committed:
Xtensa `LITBASE`: add a basic implementation

b1llow committed:
Xtensa `LITBASE`: add a integration test

b1llow committed:
Xtensa: fix cs_v6_release_guide.md

b1llow committed:
Xtensa: fix `XTENSA_OP_GROUP_MEMOPERAND`

XVilka commented:
@imbillow please rebase one more time

Rot127 commented on include/capstone/xtensa.h:
```diff
@@ -0,0 +1,181 @@
+	// generated content <XtensaGenCSFeatureEnum.inc> begin
+	// clang-format off
+
+XTENSA_FEATURE_HASDENSITY = 128,
```
Please `clang-format` the file.

Rot127 commented on cs_priv.h:
```diff
@@ -81,6 +81,7 @@ struct cs_struct {
 	const uint8_t *regsize_map;	// map to register size (x86-only for now)
 	GetRegisterAccess_t reg_access;
 	struct insn_mnem *mnem_list;	// linked list of customized instruction mnemonic
+	size_t LITBASE;
```
In the ISA `LITBASE` is defined as with `32bit`. Please use `uint32_t` here. You can cast the value when assigning it.

Rot127 commented on suite/auto-sync/src/autosync/ASUpdater.py:
```diff
@@ -60,7 +60,6 @@ def __init__(
                 USteps.INC_GEN,
                 USteps.TRANS,
                 USteps.DIFF,
-                USteps.MC,
```
Why remove this line?

b1llow replied:
It was my mistake.

Rot127 commented on suite/auto-sync/src/autosync/MCUpdater.py:
```diff
@@ -478,9 +481,13 @@ def gen_all(self):
         disas_tests = self.mc_dir.joinpath(f"Disassembler/{self.arch}")
         test_paths = [disas_tests]
+        if self.arch == "Xtensa":
+            test_paths.append(self.mc_dir.joinpath(self.arch))
```
Please add a comment that `Xtensa` only defines assembly tests.

Rot127 commented on tests/integration/test_litbase.c:
```diff
@@ -0,0 +1,77 @@
+		for (int i = 0; i < count; ++i) {
+			print_string_hex((unsigned char *)DATA,
+					 sizeof(DATA) - 1);
+			printf("\t%s\t%s\n", insn[i].mnemonic, insn[i].op_str);
```
Just printing is not enough :D
Please compare the output the expected asm text. And fail with an error message if it doesn't match.

Rot127 commented on tests/integration/CMakeLists.txt:
```diff
@@ -3,7 +3,7 @@ cmake_minimum_required(VERSION 3.15)
 if (CAPSTONE_BUILD_LEGACY_TESTS)
   enable_testing()
-  set(TEST_SOURCES test_skipdata.c test_iter.c)
+  set(TEST_SOURCES test_skipdata.c test_iter.c test_litbase.c)
```
This is no legacy test.
You can copy the code but use an `INTEGRATION_TEST_SRC` variable or similar. But they should be build by default.

Also add it as test with `add_test(NAME "integration_litbase" ...)`.

Rot127 reviewed (changes requested):
Nice, almost done!
Except the few comments. Please add `Xtensa` to the `.github/workflows/auto-sync.yml` where the other archs are.
And of course fix the CI please.
Then we are ready to merge.

b1llow committed:
Xtensa: fix

b1llow committed:
Xtensa: fix Targets.py

XVilka commented:
@imbillow few `clang-tidy` warnings left:
```
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:76:29: warning: The left operand of '==' is a garbage value due to array index out of bounds [clang-analyzer-core.UndefinedBinaryOperatorResult]
   76 |                 if (SRDecoderTable[i + 1] == RegNo) {
      |                                           ^
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:270:1: note: Control jumps to 'case 11:'  at line 270
  270 | DecodeToMCInst(decode_to_MCInst, field_from_inst, uint32_t);
      | ^
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaGenDisassemblerTables.inc:310:3: note: expanded from macro 'DecodeToMCInst'
  310 |   switch (Idx) { \
      |   ^
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:270:1: note: Calling 'DecodeSRRegisterClass'
  270 | DecodeToMCInst(decode_to_MCInst, field_from_inst, uint32_t);
      | ^
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaGenDisassemblerTables.inc:390:20: note: expanded from macro 'DecodeToMCInst'
  390 |     if (!Check(&S, DecodeSRRegisterClass(MI, tmp, Address, Decoder))) { return MCDisassembler_Fail; } \
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:72:6: note: Assuming 'RegNo' is <= 255
   72 |         if (RegNo > 255)
      |             ^~~~~~~~~~~
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:72:2: note: Taking false branch
   72 |         if (RegNo > 255)
      |         ^
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:75:2: note: Loop condition is true.  Entering loop body
   75 |         for (unsigned i = 0; i + 1 < sizeof(SRDecoderTable); i += 2) {
      |         ^
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:76:7: note: Assuming the condition is false
   76 |                 if (SRDecoderTable[i + 1] == RegNo) {
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:76:3: note: Taking false branch
   76 |                 if (SRDecoderTable[i + 1] == RegNo) {
      |                 ^
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:75:2: note: Loop condition is true.  Entering loop body
   75 |         for (unsigned i = 0; i + 1 < sizeof(SRDecoderTable); i += 2) {
      |         ^
/home/runner/work/capstone/capstone/arch/Xtensa/XtensaDisassembler.c:76:29: note: The left operand of '==' is a garbage value due to array index out of bounds
   76 |                 if (SRDecoderTable[i + 1] == RegNo) {
      |                     ~~~~~~~~~~~~~~~~~~~~~ ^
-e 

clang-tidy warnings for:  arch/Xtensa/XtensaDisassembler.c
```
Also the generation is failing:
```
Traceback (most recent call last):
  File "/home/runner/work/capstone/capstone/suite/auto-sync/./src/autosync/ASUpdater.py", line 335, in <module>
CRITICAL - Generation failed
    main()
  File "/home/runner/work/capstone/capstone/suite/auto-sync/./src/autosync/ASUpdater.py", line 331, in main
    Updater.update()
  File "/home/runner/work/capstone/capstone/suite/auto-sync/./src/autosync/ASUpdater.py", line 171, in update
    self.inc_generator.generate()
  File "/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/autosync/IncGenerator.py", line 52, in generate
    self.gen_incs()
  File "/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/autosync/IncGenerator.py", line 128, in gen_incs
    raise e
  File "/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/autosync/IncGenerator.py", line 122, in gen_incs
    subprocess.run(
  File "/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/runner/work/capstone/capstone/suite/auto-sync/vendor/llvm_root/build/bin/llvm-tblgen', '--printerLang=CCS', '--gen-disassembler', '-I', '/home/runner/work/capstone/capstone/suite/auto-sync/vendor/llvm_root/llvm/include', '-I', '/home/runner/work/capstone/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/Xtensa', '-o', '/home/runner/work/capstone/capstone/suite/auto-sync/build/llvm_c_inc/XtensaGenDisassemblerTables.inc', '/home/runner/work/capstone/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/Xtensa/Xtensa.td']' died with <Signals.SIGABRT: 6>.
```

XVilka commented:
@Rot127 according to logs, the fuzzer fails in the SystemZ, take a look please separately: https://github.com/capstone-engine/capstone/actions/runs/11043414267/job/30677644078?pr=2380#step:5:54

Rot127 committed:
Fix leaks

Rot127 committed:
Remove unnecessary new lines

Rot127 committed:
Add checks for actual buffer length before attempting reading it.

Rot127 committed:
Use isUint and isInt all over Xtensa

Rot127 committed:
Add documentation about LITBASE functionality

Rot127 committed:
Fix typo

Rot127 committed:
Replace hard with Capstone assert

Rot127 commented:
@imbillow I fixed some issue while being at it: https://github.com/imbillow/capstone/pull/1


Please check this issue here with `test_litbase`:
```
ctest --test-dir build --output-on-failure -R integration*
Internal ctest changing into directory: /home/user/repos/capstone/build
Test project /home/user/repos/capstone/build
...
    Start 11: integration_test_litbase
3/3 Test #11: integration_test_litbase .........Subprocess aborted***Exception:   0.16 sec
test_litbase: /home/user/repos/capstone/arch/Xtensa/XtensaInstPrinter.c:131: printL32RTarget: Assertion `(Value >= -262144 && Value <= -4) && "Invalid argument, value must be in ranges [-262144,-4]"' failed.
```

Also, it is weird. But in the CI it builds Xtensa without my edits. But it shouldn't because those macros don't exist. Or did you add them and I just overlooked them?

b1llow commented:
The clang-tidy warnings are a false positive, because `SRDecoderTable[i + 1]` does not cause the array to go out of bounds if `i + 1 < sizeof(SRDecoderTable)` is satisfied.

b1llow committed:
Xtensa: fix arch_config.json

b1llow commented:
`Auto-Sync/check` fix in https://github.com/capstone-engine/llvm-capstone/pull/61

Rot127 commented on arch/Xtensa/XtensaDisassembler.c:
```diff
@@ -0,0 +1,298 @@
+	if (RegNo > 255)
+		return MCDisassembler_Fail;
+
+	for (unsigned i = 0; i + 1 < sizeof(SRDecoderTable); i += 2) {
```
> The clang-tidy warnings are a false positive, because SRDecoderTable[i + 1] does not cause the array to go out of bounds if i + 1 < sizeof(SRDecoderTable) is satisfied.

We are not in C++ realm here unfortunately. `sizeof` returns the number of bytes. Not the number of elements.
Please use the macro `ARR_SIZE()` for this.

Rot127 commented on arch/Xtensa/XtensaDisassembler.c:
```diff
@@ -0,0 +1,298 @@
+static DecodeStatus DecodeARRegisterClass(MCInst *Inst, uint64_t RegNo,
+					  uint64_t Address, const void *Decoder)
+{
+	if (RegNo >= sizeof(ARDecoderTable))
```
Also here

b1llow committed:
Xtensa: fix

Rot127 reviewed (approved):
@kabeor Your turn

kabeor reviewed (approved):
WoW, really cool, thanks!
